### DISCUSSION TITLE
Would  async with ASGI be helpful here?

### DISCUSSION THREAD

**Discussion by antont** [snarfed/bridgy-fed#702] [Ideas]
Hi, just wondering, are you happy with how Flask works here, or would it be useful to switch to ASGI and use e.g. Starlette?

I've ported some things from WSGI -> ASGI, from Flask to Starlette (with FastAPI), and have been happy with the results, also when running on the appengine, using gunicorn with uvicorn.workers.UvicornWorker

**Comment by snarfed** [comment]
I'd love to! asyncio and ASGI are great. It would be a huge project though, and wouldn't solve any real Bridgy Fed problems right now that I'm aware of. Is there some concrete pain point here that you think ASGI would help with?

I'd also need async support in all of the I/O libraries I use, notably requests and the Google Cloud API libraries. Afaict library support tends to be the long pole for adopting asyncio in general.

**Reply by antont** [reply]
No, I don't have a concrete pain point, was curious if you have encountered any.

It is a substantial effort, yes. I did crunch it in once in a about 2 full working days for a maybe similar sized project.

Also, it can be done in parts, starting from async entry points and porting some of the implementation to async, while rest remain sync, because you can always call sync code from async. It just blocks then and you loose some of the benefit, but that does not make it worse than the fully sync version.

I'm interested in helping out here in general, not sure if will actually have time but maybe some, but will keep an eye on actually immediately helpful issues first. Perhaps at some later time would be fun to try out an async port.

Oh and in the work thing where we use async we also run on Google cloud and use some of the async libs for it, basically there is just storage though, and then 3rd party impls for others but we're not using those. Earlier we used Postgres and now Firebase, where async is supported in the official SDK, so we benefit there.

But it's useful also when calling external services, which the bridge does a lot by defition, so using httpx or aiohttp (we use that now) would be nice, to be able to serve multiple requests concurrently in a single process.

**Reply by snarfed** [reply]
Sure! Any and all help here is welcome, feel free to check out the developer docs and open issues and ask me any q's.

(Btw we currently use threaded gunicorn workers, so we definitely don't block entire worker processes on individual network requests.)
